crates/red_knot_python_semantic/src/types.rs

```diff
@@ -245,46 +263,77 @@ fn definition_expression_ty<'db>(
-                .peekable();
-
-            let binding_ty = binding_ty(db, binding);
-            if constraint_tys.peek().is_some() {
-                constraint_tys
-                    .fold(
-                        IntersectionBuilder::new(db).add_positive(binding_ty),
-                        IntersectionBuilder::add_positive,
-                    )
-                    .build()
-            } else {
-                binding_ty
+) -> Symbol<'db> {
+    let mut is_unbound_visible = false;
+    let mut types = vec![];
+
+    for BindingWithConstraints {
+        binding,
+        constraints,
+        all_constraints,
+        visibility_constraints,
+        visibility_constraint,
+    } in bindings_with_constraints
+    {
+        let static_visibility =
+            visibility_constraints.analyze(db, all_constraints, visibility_constraint);
+
+        let Some(binding) = binding else {
+            if !static_visibility.is_always_false() {
+                // TODO: also distinguish between always-true and ambiguous?
+                is_unbound_visible = true;
             }
-        },
-    );
+            continue;
+        };
 
-    if let Some(first) = def_types.next() {
-        if let Some(second) = def_types.next() {
-            Some(UnionType::from_elements(
-                db,
-                [first, second].into_iter().chain(def_types),
-            ))
+        if static_visibility.is_always_false() {
+            continue;
+        }
+
+        let mut constraint_tys = constraints
+            .filter_map(|constraint| narrowing_constraint(db, constraint, binding))
+            .peekable();
+
+        let binding_ty = binding_ty(db, binding);
+        let ty = if constraint_tys.peek().is_some() {
+            let intersection_ty = constraint_tys
+                .fold(
+                    IntersectionBuilder::new(db).add_positive(binding_ty),
+                    IntersectionBuilder::add_positive,
+                )
+                .build();
+            intersection_ty
+        } else {
+            binding_ty
+        };
+
+        types.push(ty);
+    }
+
+    let mut types = types.into_iter();
+
+    if let Some(first) = types.next() {
+        let boundness = if is_unbound_visible {
+            Boundness::PossiblyUnbound
         } else {
-            Some(first)
+            Boundness::Bound
+        };
+
+        if let Some(second) = types.next() {
+            Symbol::Type(
+                UnionType::from_elements(db, [first, second].into_iter().chain(types)),
+                boundness,
+            )
+        } else {
+            Symbol::Type(first, boundness)
         }
     } else {
-        None
+        Symbol::Unbound
     }
 }
 
 /// The result of looking up a declared type from declarations; see [`declarations_ty`].
-type DeclaredTypeResult<'db> = Result<Type<'db>, (Type<'db>, Box<[Type<'db>]>)>;
+type DeclaredTypeResult<'db> = Result<Symbol<'db>, (Type<'db>, Box<[Type<'db>]>)>;
```

Comment:
Hmm, I see that by using `Symbol` here, which includes "Boundness", we give up on distinguishing boundness (applies to bindings) from declaredness (applies to declarations). Maybe that's fine as temporary state in this PR, something to consider as follow up, along with the whole issue of declaredness vs boundness.

Reply:
Yes. I am using "boundness" as "declaredness" in some cases now.

I have the feeling that it will be easier to solve those declaredness-vs-boundness issues you mentioned with the changes on this branch, as we simplify the boundness/undeclaredness logic quite a bit. But let me know if we should do something here and now.

Reply:
No, I think this is fine for now, and we should have a follow up on how we treat boundness vs declaredness.